###Survival Analysis

* This notebook is about the application of Survival Analysis in cricket - to analyze the career lengths of players.

* Survival Analysis has been described in length by various resources, the following are the major soruces I used to
understand the concept:
    * Allen B. Downy's book on exploratory data analysis in Python includes a great chapter on survival curves, hazard functions, Kaplan–Meier estimators etc. 
    
    http://greenteapress.com/thinkstats2/thinkstats2.pdf 
    
    * Econometrics Academy's notes on survival analysis http://sites.google.com/site/econometricsacademy/econometrics-models/survival-analysis
     
    * Cam Davidson-Pilon's documentation on the lifelines python library on survival analysis.
    
         http://lifelines.readthedocs.org/en/latest/Quickstart.html

* Survival Analyis is used in areas where the time duration of a sample of observations is analysed until an event of death occurs.
* This has great use in mechanical engineering where the lifetime of a tool/product is analyzed, medical sciences where 
the lifetime of cancer patients is analyzed etc.
* This is an attempt to extend this statistical concept into the field of cricket - to analyze the career lengths of players.

* The event of death in this case is the event when players retire from active cricket.
* I have tried to analyze all the players who have played ODI cricket. 
* The reason Test cricket wasn't chosen is due to the fact there is too much noise in the data due to the careers of the players marred due to the World Wars, Apartheid Crisis, Kerry Packer's cricket series etc.

- There isn't any readily available data when it comes to cricket yet - ESPNCricinfo still doesn't provide an API to use its StatsGuru database machine - so I had to scrape the data from the Statsguru webpages to acquire the data.

In [4]:
%matplotlib inline
from bs4 import BeautifulSoup
import requests
import pandas as pd
import lifelines
import matplotlib
import numpy as np
import matplotlib.pyplot as plt
matplotlib.style.use('ggplot')

#### Scraping
The following method scrapes the required data from the webpages. 

In [14]:
statsguru_query_url = 'http://stats.espncricinfo.com/ci/engine/stats/index.html?class=2;filter=advanced;orderby=runs;page=1;size=200;template=results;type=batting'

In [15]:
def scrape_data(page_count):
    """
    scrapes the required data present in the form of a table from the given url
    :param url: page of the ESPNCricinfo Stats URL query
    :return: the raw unicode text
    """
    url = statsguru_query_url
    complete_url = url.partition('page=1')[0] + "page=" + str(page_count) + url.partition('page=1')[-1]
    r = requests.get(complete_url)
    data = r.text
    soup = BeautifulSoup(data)
    table = soup.find_all('table')
    return table[2].text

#### Data Cleaning
Once the data is scraped, it has to be cleaned - stripping of the whitespaces and other noisy data to get it into a proper structure

In [16]:
def clean_data(text):
    """
    cleans the data removing all whitespaces and other data
    :param text: table text of each page
    :return: cleaned text  
    """
    text = text.split('\n')
    text.remove(u'Overall figures')
    return text

#### Data Transformation
The data is now transformed accordingly to make get the data into a structure with which it can effectively be modelled.
A series of methods and transformations are done to make a dataframe, fit for modelling.

In [73]:
def create_data():
    """
    scrapes the data, cleans it and transforms the data to load into a pandas dataframe
    """
    page_count = 12

    df = pd.DataFrame()

    get_list_columns = lambda text, start_index, end_index: [str(unicode_text)
                                                                for unicode_text in text][start_index:end_index]

    get_data_rows = lambda text, start_index: text[start_index:]

    remove_all_occurences = lambda data, item: [x for x in data if x != item]

    get_list_rows = lambda data: [data[index: index + 13] for index, row in enumerate(data) if index % 13 == 0]

    for page in range(1, page_count + 1):
        raw_text = scrape_data(page)
        clean_text = clean_data(raw_text)
        list_columns = get_list_columns(clean_text, 3, 16)
        data_rows = get_data_rows(clean_text, 16)
        data_rows = remove_all_occurences(data_rows, u'')
        list_rows = get_list_rows(data_rows)
        df_new = pd.DataFrame(list_rows, columns=list_columns)
        if len(df) == 0:
            df = df_new
        else:
            df = pd.concat([df, df_new])
    return df

In [21]:
df_full = create_data()
df_full = df_full.reset_index(drop=True)

* The scraped data is stored in the form of a pandas dataframe.

In [25]:
len(df_full)

2244

* There are totally 2244 players who have played ODI cricket since its inception in the 1970s.

* The following displays the first five rows of the data.
* The data is by default sorted by the amount of runs scored.

In [23]:
df_full.head()

Player       Span  Mat Inns  NO   Runs    HS    Ave  \
0         SR Tendulkar (India)  1989-2012  463  452  41  18426  200*  44.83   
1  KC Sangakkara (Asia/ICC/SL)  2000-2015  404  380  41  14234   169  41.98   
2         RT Ponting (Aus/ICC)  1995-2012  375  365  39  13704   164  42.03   
3      ST Jayasuriya (Asia/SL)  1989-2011  445  433  18  13430   189  32.36   
4   DPMD Jayawardene (Asia/SL)  1998-2015  448  418  39  12650   144  33.37   

      BF     SR 100  50   0  
0  21367  86.23  49  96  20  
1  18048  78.86  25  93  15  
2  17046  80.39  30  82  20  
3  14725  91.20  28  68  34  
4  16020  78.96  19  77  28

Renaming the columns to access them easier

In [28]:
df_full.columns = ['player', 'span', 'mat', 'inns', 'not_outs', 'runs', 'high_score', 'ave', 'bf', 'sr', 'n_100', 'n_50', 'n_0']

In [29]:
df_span = df_full[['player','span']]

In [30]:
calc_career_length = lambda span: [ int(each_span.partition('-')[-1]) - int(each_span.partition('-')[0]) for each_span in span]
df_span['career_length'] = calc_career_length(df_span.span)

/Users/kramaswamy/anaconda/lib/python2.7/site-packages/IPython/kernel/__main__.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  from IPython.kernel.zmq import kernelapp as app


In [40]:
df_span['censor'] = [1 for _ in range(0,len(df_span))]

/Users/kramaswamy/anaconda/lib/python2.7/site-packages/IPython/kernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


In [54]:
df_span['career_start_date'] = [int(span.partition('-')[0]) for span in df_span.span]
df_span['career_end_date'] = [int(span.partition('-')[-1]) for span in df_span.span]

/Users/kramaswamy/anaconda/lib/python2.7/site-packages/IPython/kernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':
/Users/kramaswamy/anaconda/lib/python2.7/site-packages/IPython/kernel/__main__.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  from IPython.kernel.zmq import kernelapp as app


In [71]:
df_span

player       span  career_length  censor  \
0             SR Tendulkar (India)  1989-2012             23       1   
1      KC Sangakkara (Asia/ICC/SL)  2000-2015             15       1   
2             RT Ponting (Aus/ICC)  1995-2012             17       1   
3          ST Jayasuriya (Asia/SL)  1989-2011             22       1   
4       DPMD Jayawardene (Asia/SL)  1998-2015             17       1   
5        Inzamam-ul-Haq (Asia/Pak)  1991-2007             16       1   
6           JH Kallis (Afr/ICC/SA)  1996-2014             18       1   
7          SC Ganguly (Asia/India)  1992-2007             15       1   
8        R Dravid (Asia/ICC/India)  1996-2011             15       1   
9                 BC Lara (ICC/WI)  1990-2007             17       1   
10                 TM Dilshan (SL)  1999-2015             16       1   
11      Mohammad Yousuf (Asia/Pak)  1998-2010             12       1   
12          AC Gilchrist (Aus/ICC)  1996-2008             12       1   
13            M Azharuddin (India)  1985-2000             15       1   
14                PA de Silva (SL)  1984-2003             19       1   
15               CH Gayle (ICC/WI)  1999-2015             16       1   
16               Saeed Anwar (Pak)  1989-2003             14       1   
17              S Chanderpaul (WI)  1994-2011             17       1   
18                  DL Haynes (WI)  1978-1994             16       1   
19           MS Dhoni (Asia/India)  2004-2015             11       1   
20                MS Atapattu (SL)  1990-2007             17       1   
21                  ME Waugh (Aus)  1988-2002             14       1   
22       Yuvraj Singh (Asia/India)  2000-2013             13       1   
23       V Sehwag (Asia/ICC/India)  1999-2013             14       1   
24                   HH Gibbs (SA)  1996-2010             14       1   
25    Shahid Afridi (Asia/ICC/Pak)  1996-2015             19       1   
26         AB de Villiers (Afr/SA)  2005-2015             10       1   
27             SP Fleming (ICC/NZ)  1994-2007             13       1   
28                 MJ Clarke (Aus)  2003-2015             12       1   
29                  SR Waugh (Aus)  1986-2002             16       1   
30                A Ranatunga (SL)  1982-1999             17       1   
31             Javed Miandad (Pak)  1975-1996             21       1   
32               Younis Khan (Pak)  2000-2015             15       1   
33              Saleem Malik (Pak)  1982-1999             17       1   
34                   NJ Astle (NZ)  1995-2007             12       1   
35               GC Smith (Afr/SA)  2002-2013             11       1   
36                  MG Bevan (Aus)  1994-2004             10       1   
37                  G Kirsten (SA)  1993-2003             10       1   
38                  A Flower (Zim)  1992-2003             11       1   
39               IVA Richards (WI)  1975-1991             16       1   
40                 V Kohli (India)  2008-2015              7       1   
41                 GW Flower (Zim)  1992-2010             18       1   
42                Ijaz Ahmed (Pak)  1986-2000             14       1   
43                 AR Border (Aus)  1979-1994             15       1   
44              RB Richardson (WI)  1983-1996             13       1   
45             ML Hayden (Aus/ICC)  1993-2008             15       1   
46                  DM Jones (Aus)  1984-1994             10       1   
47                   DC Boon (Aus)  1984-1995             11       1   
48                  JN Rhodes (SA)  1992-2003             11       1   
49                    HM Amla (SA)  2008-2015              7       1   
50                BB McCullum (NZ)  2002-2015             13       1   
51               Rameez Raja (Pak)  1985-1997             12       1   
52              Shoaib Malik (Pak)  1999-2015             16       1   
53                  RR Sarwan (WI)  2000-2013             13       1   
54                  CL Hooper (WI)  1987-2003             16       1